### PR TITLE
Scripts/SholazarBasin: Make quest "Engineering a Disaster" more retail-like

### PR DESCRIPTION
**Pull request by Carbenium**
**Changes proposed:**

- Use correct spells and add the appropriate spell scripts
- Use correct flame GOs
- Fix emotes of the escortee

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Tests performed:** build + in-game

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by offl** [comment]
Was going to convert it to SAI :D

**Comment by Carbenium** [comment]
@offl: Please go ahead! That should work now with the spell scripts implemented.

File: src/server/scripts/Northrend/zone_sholazar_basin.cpp [hunk 1]
```diff
@@ -50,109 +51,147 @@ enum EngineerHelice
...
-                else
-                    m_uiChatTimer -= uiDiff;
-            }
+                Position Pos;
+                QuaternionData Quat;
+            } Flames[] =
+                    {
+                            {{ 5921.5337f, 5371.778f,  -96.338776f, 0.f }, { 0.f, 0.f, 0.6883545f,   0.72537446f }},
+                            {{ 5919.418f,  5366.918f,  -96.091125f, 0.f }, { 0.f, 0.f, 0.60876083f,  0.7933538f }},
+                            {{ 5920.2563f, 5372.4136f, -98.85827f,  0.f }, { 0.f, 0.f, -0.32556725f, 0.94551885f }}
```
Comment (Carbenium) [inline src/server/scripts/Northrend/zone_sholazar_basin.cpp:138]: I'm not happy with spawning the flames like this. Investigating a cleaner way.

Comment (offl) [reply src/server/scripts/Northrend/zone_sholazar_basin.cpp:138]: You can spawn them in DB with negative respawn time and respawn in script

**Comment by Carbenium** [comment]
@Treeston could you take a look at the flame spawning? First steps with the spawn system... Is there no better way to do a timed despawn?

**Comment by illfated** [comment]
When you get as far as to finalizing the PR, I recommend applying TC SQL codestyle to the `2020_07_xx_xx_world.sql` file (back ticks around the table names and column names and so on).

**Comment by Carbenium** [comment]
Blocked by #25179

**Comment by offl** [comment]
Better to merge first version without involving spawn groups and summon_groups just to merge this

**Comment by Carbenium** [comment]
Yeah, you might be right... I don't have the capacity right now to implement the GO summon groups.

I'll try to rebase this on the weekend so we can merge it as it is now.

**Comment by illfated** [comment]
By the way, I have edited the SQL file content (as of the time of this comment) to make it easy to copy & paste it back to your PR file:
```sql
-- NPC Engineer Helice
UPDATE `creature_text` SET `emote`=  5 WHERE `CreatureID`= 28787 AND `GroupID`= 0;
UPDATE `creature_text` SET `emote`= 25 WHERE `CreatureID`= 28787 AND `GroupID`= 1;
UPDATE `creature_text` SET `emote`= 11 WHERE `CreatureID`= 28787 AND `GroupID`= 3;
UPDATE `creature_text` SET `emote`= 25 WHERE `CreatureID`= 28787 AND `GroupID`= 4;
UPDATE `creature_text` SET `emote`=  5 WHERE `CreatureID`= 28787 AND `GroupID`= 5;
UPDATE `creature_text` SET `emote`=  4 WHERE `CreatureID`= 28787 AND `GroupID`= 6;

DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`= 13 AND `SourceEntry` IN (52369, 52371);
INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorType`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
(13, 1, 52369, 0, 0, 31, 0, 3, 28789, 0, 0, 0, 0, '', "Spell 'Detonate Explosives' targets trigger"),
(13, 1, 52371, 0, 0, 31, 0, 3, 28789, 0, 0, 0, 0, '', "Spell 'Detonate Explosives' targets trigger");

DELETE FROM `spell_script_names` WHERE `spell_id` IN (52369, 52371);
INSERT INTO `spell_script_names` (`spell_id`,`ScriptName`) VALUES
(52369, 'spell_q12688_detonate_1'),
(52371, 'spell_q12688_detonate_2');

-- Flames
DELETE FROM `gameobject` WHERE `guid` IN (9577, 9578, 9579, 9598, 9599, 9600);
INSERT INTO `gameobject` (`guid`,`id`,`map`,`zoneId`,`areaId`,`spawnMask`,`phaseMask`,`position_x`,`position_y`,`position_z`,`orientation`,`rotation0`,`rotation1`,`rotation2`,`rotation3`,`spawntimesecs`,`animprogress`,`state`,`ScriptName`,`VerifiedBuild`) VALUES
(9577, 182071, 571, 0, 0, 1, 1, 5921.53, 5371.77, -96.3387, 0, 0, 0,  0.688354,    0.725374,   0, 0, 1, '', 0),
(9578, 182071, 571, 0, 0, 1, 1, 5919.42, 5366.92, -96.0911, 0, 0, 0,  0.608761,    0.793354,   0, 0, 1, '', 0),
(9579, 182071, 571, 0, 0, 1, 1, 5920.25, 5372.41, -98.8582, 0, 0, 0, -0.325567,    0.945519,   0, 0, 1, '', 0),
(9598, 182071, 571, 0, 0, 1, 1, 5888.68, 5377.42, -92.6171, 0, 0, 0,  0.9612608,   0.2756405,  0, 0, 1, '', 0),
(9599, 182071, 571, 0, 0, 1, 1, 5886.48, 5379.05, -92.9799, 0, 0, 0,  0.41469288,  0.90996146, 0, 0, 1, '', 0),
(9600, 182071, 571, 0, 0, 1, 1, 5887.20, 5381.18, -93.5000, 0, 0, 0,  0.061048508, 0.9981348,  0, 0, 1, '', 0);

DELETE FROM `spawn_group_template` WHERE `groupId` IN (67, 68);
INSERT INTO `spawn_group_template` (`groupId`,`groupName`,`groupFlags`) VALUES
(67, 'Quest "Engineering a Disaster" (12688) - Spell Detonate (52369)', 4),
(68, 'Quest "Engineering a Disaster" (12688) - Spell Detonate (52371)', 4);

DELETE FROM `spawn_group` WHERE `groupId` IN (67, 68);
INSERT INTO `spawn_group` (`groupId`,`spawnType`,`spawnId`) VALUES
(67, 1, 9577),
(67, 1, 9578),
(67, 1, 9579),
(68, 1, 9598),
(68, 1, 9599),
(68, 1, 9600);
```

**Comment by jackpoz** [comment]
@Carbenium can some bits of this PR be merged even if not complete ? Otherwise I will close this as abandoned

**Comment by Carbenium** [comment]
Rebased and re-tested

**Comment by offl** [comment]
So with spawn groups?

**Comment by Carbenium** [comment]
Yeah, spawn groups and timed despawns from the script for the time being.

File: src/server/scripts/Northrend/zone_sholazar_basin.cpp [hunk 2]
```diff
@@ -50,109 +52,129 @@ enum EngineerHelice
...
-                    m_uiChatTimer = 12000;
-                }
-                else
-                    m_uiChatTimer -= uiDiff;
-            }
+            std::vector<WorldObject*> flames;
+            target->GetMap()->SpawnGroupSpawn(SPAWN_GROUP_FLAMES, false, false, &flames);
+            target->GetMap()->SetSpawnGroupInactive(SPAWN_GROUP_FLAMES);
+            for (WorldObject* flame : flames)
+                dynamic_cast<GameObject*>(flame)->DespawnOrUnsummon(20s);
```
Comment (jackpoz) [inline src/server/scripts/Northrend/zone_sholazar_basin.cpp:133]: @Carbenium coverity reported this as static analysis issue as dynamic_cast can return null. Either null-check ToGameObject() result or assert (null check ToGameObject() should be better as there is little gain on crashing the whole server with an assert)

I can do it in the next days if you are busy.

Coverity issues 1457830 and 1457831 for both dynamic_cast lines in this file (line 172 too)

**Pull request merged**